AK/InsertionSort.h

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2022, Marc Luqué <marc.luque@outlook.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Concepts.h>
+#include <AK/StdLibExtras.h>
+
+namespace AK {
+
+// Standard Insertion Sort, with `end` inclusive!
+template<typename Collection, typename Comparator, typename T = decltype(declval<Collection>()[declval<int>()])>
+void insertion_sort(Collection& col, int start, int end, Comparator comparator) requires(Indexable<Collection, T>)
+{
+    for (int i = start; i <= end; ++i) {
+        for (int j = i; j > 0 && comparator(col[j], col[j - 1]); --j)
+            swap(col[j], col[j - 1]);
+    }
+}
+
+// `end` is exclusive, as usual with iterators.
+template<typename Iterator, typename Comparator, typename T = decltype(*declval<Iterator>())>
+void insertion_sort(Iterator start, Iterator end, Comparator comparator) requires(IteratorType<Iterator, T>)
+{
+    for (Iterator i = start; i < end; ++i) {
+        for (Iterator j = i; j > start && comparator(*j, *(j - 1)); --j)
```

Comment:
This requires less-than/greater-than on the iterators, that's not implemented in every iterator and it's not actually possible to do so for every iterator either.

Reply:
But then we just can't use insertion sort, it relies on that existing... another requires clause maybe? Other ideas?

Reply:
It's not really about adding more constraints, doing that would at best make errors better.
unfortunately this is library code and is supposed to be generic enough, and when you go out of your way to add elaborate constraints, they should at least match your real requirements, why not just do what quick_sort does and not care? currently you have these requirements on the iterator:
- `--it`
- iterator copy
- `it - int`
- `it0 > it1`
- `it0 < it1`
- `comparator(*it0, *it1)` should be valid

certainly you _could_ write this in a way that would only require iterator copy, `++it`, and `it0 == it1` and the comparator requirement as an implicit one by going forwards from `j = start -> i` and keeping a `previous_j` copy, something like:
```c++
    for (auto i = start; i != end; ++i) {
        auto j = start;
        auto previous_j = j;
        ++j;
        for (; j != i && !comparator(*j, *previous_j); ++j, ++previous_j)
            swap(*j, *previous_j);
    }
```
(I'm not sure about the semantics here, but I'm certain you can figure it out)

As an example for an "iterator" that's more like a torture test, see:
```c++
struct ArbitraryIterable {
    struct EndIterator { };

    struct ArbitraryIterator {
        ArbitraryIterator(int& v)
            : value(v)
        {
        }

        bool operator==(EndIterator const&) const { return times_generated >= 5; }

        bool operator==(ArbitraryIterator const& other) const
        {
            return &value == &other.value && times_generated == other.times_generated;
        }

        int& operator*()
        {
            ++value;
            return value;
        }

        ArbitraryIterator& operator++()
        {
            ++times_generated;
            return *this;
        }

        int& value;
        int times_generated { 0 };
    };

    ArbitraryIterator begin() { return value; }
    EndIterator end() const { return {}; }
    int value { 1 };
};
```
Does this make sense? yes, but would anyone be silly enough to do this? probably not directly, but it's quite possible to have an iterator that has at least one of the properties of this thing.

that's generally why the iterator protocol is a pain and you shouldn't assume random operators exist (such things do exist, see the HashTable iterator, or the CircularQueue iterator, or even the IntrusiveList iterator for examples)

Reply:
(see https://serenity.godbolt.org/z/jETvG43qd)

Reply:
The reason we go backwards in the inner loop is due to how this insertion sort implementation inserts small elements from the top downwards, progressively, until the correct place is found. Your simple fix won't work because we need to actually loop downwards for shifting smaller elements down. Now, you can of course do it the other way around and shift larger elements upwards progressively. But then that requires the outer loop bound to progressively decrease, as it needs to keep track of what was already sorted at the top. In essence, both loops have to go in opposite directions.